### ISSUE TITLE
Hangs with sending/receiving

### ISSUE THREAD

**Issue by jakirkham** [rapidsai/ucx-py#150]
Have seen some hangs when trying to send and receive data. In particular when running `pytest tests/test_send_recv_obj.py::test_send_recv_bytes`, it will run the first time and hang the second time.

**Comment by jakirkham** [comment]
@pentschev, suggested setting the following environment variables.

```
UCX_RNDV_SCHEME=put_zcopy
UCX_MEMTYPE_CACHE=n
UCX_TLS=rc,cuda_copy
```

This seems to fix the issue for me.

**Comment by quasiben** [comment]
Let's move these env vars into the tests and print them out during execution.  I think we've cycled through this problem a number of times

**Comment by jakirkham** [comment]
If these environment variables need to be set generally regardless of setup, we might consider adding an [activation script]( https://conda-forge.org/docs/maintainer/adding_pkgs.html#activate-scripts ) to the `ucx` package that sets these for the user. Though am not clear on whether they are needed generally or just the DGXs we are testing. @Akshay-Venkatesh, do you have thoughts on this?

**Comment by quasiben** [comment]
@jakirkham there is some subtlety here.  For example, `UCX_TLS=rc,cuda_copy`, i believe, means use infiniband.  if you add `cuda_ipc` that means also use nvlink.  @Akshay-Venkatesh can you comment on what happens with UCX if we use `cuda_ipc` but don't have NVLINK hardware

**Comment by pentschev** [comment]
I'm not sure I like the idea of occluding the setup of these variables. I think for newcomers it will make things more difficult to debug, e.g., tests pass but a custom user script will hang when they're not set explicitly.

**Comment by Akshay-Venkatesh** [comment]
cuda_ipc works over pcie for certain configurations so that gets used. In general, cuda_ipc shouldn't be removed from ucx_tls. 

Can you share the repro details @jakirkham?

**Comment by jakirkham** [comment]
Sure. Here you go. Afraid there is not too much more to add than the OP. So this is mainly useful for letting you recreate the environment I have. Environment variables listed above fix the hang (see comment in reproducer for where).

<details>
<summary>Reproducer:</summary> 

```
$ pytest tests/test_send_recv_obj.py::test_send_recv_bytes  # passes
$ pytest tests/test_send_recv_obj.py::test_send_recv_bytes  # hangs indefinitely without output
```

</details>

<details>
<summary>Spec file for environment:</summary>

Use with `conda create --name myenv --file spec-file.txt` as [explained here]( https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#building-identical-conda-environments ).

```
# This file may be used to create an environment using:
# $ conda create --name <env> --file <this file>
# platform: linux-64
@EXPLICIT
https://repo.anaconda.com/pkgs/main/linux-64/_libgcc_mutex-0.1-main.tar.bz2
https://conda.anaconda.org/jakirkham/label/ucx/linux-64/ucx-proc-1.0.0-gpu.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/ca-certificates-2019.6.16-hecc5488_0.tar.bz2
https://repo.anaconda.com/pkgs/main/linux-64/cudatoolkit-9.2-0.tar.bz2
https://repo.anaconda.com/pkgs/main/linux-64/libgcc-ng-9.1.0-hdf63c60_0.tar.bz2
https://repo.anaconda.com/pkgs/main/linux-64/libgfortran-ng-7.3.0-hdf63c60_0.tar.bz2
https://repo.anaconda.com/pkgs/main/linux-64/libstdcxx-ng-9.1.0-hdf63c60_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/bzip2-1.0.8-h516909a_0.tar.bz2
https://repo.anaconda.com/pkgs/main/linux-64/cudnn-7.6.0-cuda9.2_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/jpeg-9c-h14c3975_1001.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libffi-3.2.1-he1b5a44_1006.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libopenblas-0.3.7-h6e990d7_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/lz4-c-1.8.3-he1b5a44_1001.tar.bz2
https://conda.anaconda.org/nvidia/linux-64/nccl-2.4.6.1-cuda9.2_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/ncurses-6.1-hf484d3e_1002.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/openssl-1.1.1c-h516909a_0.tar.bz2
https://conda.anaconda.org/jakirkham/label/ucx/linux-64/ucx-1.7.0.dev1+g2d85597-h934ff1e_1_gpu.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/xz-5.2.4-h14c3975_1001.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/yaml-0.1.7-h14c3975_1001.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/zlib-1.2.11-h516909a_1005.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libblas-3.8.0-12_openblas.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libpng-1.6.37-hed695b0_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/readline-7.0-hf8c457e_1001.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/tk-8.6.9-hed695b0_1002.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/zstd-1.4.0-h3b9ef0a_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/freetype-2.10.0-he983fc9_1.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libcblas-3.8.0-12_openblas.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/liblapack-3.8.0-12_openblas.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/libtiff-4.0.10-h57b8799_1003.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/sqlite-3.28.0-h8b20d00_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/python-3.7.3-h5b0a415_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/atomicwrites-1.3.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/attrs-19.1.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/backcall-0.1.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/certifi-2019.6.16-py37_1.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/click-7.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/cloudpickle-1.2.1-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/dask-core-2.0.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/decorator-4.4.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/fastrlock-0.4-py37hf484d3e_1000.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/heapdict-1.0.0-py37_1000.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/ipython_genutils-0.2.0-py_1.tar.bz2
https://conda.anaconda.org/numba/linux-64/llvmlite-0.29.0-py37hf484d3e_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/locket-0.2.0-py_2.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/markupsafe-1.1.1-py37h14c3975_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/more-itertools-7.2.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/msgpack-python-0.6.1-py37h6bb024c_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/numpy-1.16.4-py37h95a1406_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/olefile-0.46-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/parso-0.5.1-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pickleshare-0.7.5-py37_1000.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/psutil-5.6.3-py37h516909a_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/ptyprocess-0.6.0-py_1001.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/py-1.8.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/pyparsing-2.4.2-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/pytz-2019.2-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pyyaml-5.1.2-py37h516909a_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/six-1.12.0-py37_1000.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/sortedcontainers-2.1.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/tblib-1.4.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/toolz-0.10.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/tornado-6.0.3-py37h516909a_0.tar.bz2
https://conda.anaconda.org/jakirkham/label/ucx/linux-64/ucx-py-0.1.0.dev1+g7d37e7e-py37h934ff1e_1_gpu.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/wcwidth-0.1.7-py_1.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/zipp-0.5.2-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/cytoolz-0.10.0-py37h516909a_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/importlib_metadata-0.18-py37_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/jedi-0.15.1-py37_0.tar.bz2
https://conda.anaconda.org/numba/linux-64/numba-0.44.1-np116py37hf484d3e_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/packaging-19.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/partd-1.0.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pexpect-4.7.0-py37_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pillow-6.1.0-py37h6b7be26_1.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/python-dateutil-2.8.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/setuptools-41.0.1-py37_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/traitlets-4.3.2-py37_1000.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/zict-1.0.0-py_0.tar.bz2
https://conda.anaconda.org/jakirkham/label/cupy/linux-64/cupy-6.2.0-py37h6adefed_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/distributed-2.0.1-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/jinja2-2.10.1-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pandas-0.25.0-py37hb3f55d8_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/pluggy-0.12.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/pygments-2.4.2-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/wheel-0.33.4-py37_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/bokeh-1.3.4-py37_0.tar.bz2
https://conda.anaconda.org/rapidsai/linux-64/dask-cuda-0.8.0-py37_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pip-19.2.2-py37_0.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/prompt_toolkit-2.0.9-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pytest-5.0.1-py37_1.tar.bz2
https://conda.anaconda.org/conda-forge/noarch/dask-2.0.0-py_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/ipython-7.5.0-py37h24bf2e0_0.tar.bz2
https://conda.anaconda.org/conda-forge/linux-64/pytest-asyncio-0.10.0-py37_1000.tar.bz2
```

</details>

**Comment by quasiben** [comment]
@Akshay-Venkatesh Thanks!

> cuda_ipc works over pcie for certain configurations so that gets used. In general, cuda_ipc shouldn't be removed from ucx_tls.

I'll amend tests appropriately

**Comment by quasiben** [comment]
resolved in #152
